modules/core/src/main/java/org/apache/ignite/internal/binary/GridBinaryMarshaller.java

```diff
@@ -150,9 +150,16 @@
     /** Time array. */
     public static final byte TIME_ARR = 37;
 
-    /** Binary enum */
+    /** Binary enum. */
     public static final byte BINARY_ENUM = 38;
 
+    /**
+     * Binary wrapper for {@code Object[]}.
+     * This wrapper used to store array component type id during serde process.
+     * @see BinaryUtils#doReadObjectArrayWrapper(BinaryInputStream, BinaryContext, ClassLoader, BinaryReaderHandlesHolder, boolean, boolean)
+     */
+    public static final byte OBJ_ARR_WRAPPER = 39;
```

Comment:
Do we really need a dedicated type for the wrapper? The wrapper is only a temporary internal representation of the array after unmarshalling, if we marshall it again we can use the same OBJ_ARR type.

Reply:
Yes. It passed through the async compute call in `GridServiceProxy`.
So ser-deser happens in a different thread. If we don't use dedicated typeid we lose array information.
```
                        // Execute service remotely.
                        return ctx.closure().callAsyncNoFailover(
                            GridClosureCallMode.BROADCAST,
                            new ServiceProxyCallable(methodName(mtd), name, mtd.getParameterTypes(), args, BinaryUtils.USE_ARRAY_WRAPPER.get()),
                            Collections.singleton(node),
                            false,
                            waitTimeout,
                            true).get();

``

Reply:
Agree with Alex, it would be great to avoid a new system type.
Otherwise we'll have to add the support for this type to .NET and C++, right?

Reply:
This type exists only during transfer from a caller to a callee.
.Net and C++ receives typed array.